### PR TITLE
Added Python typechecking for mypy for `src/paths` `tests/tests_paths`

### PR DESCRIPTION
**Pull request by abhinavGupta16**
### Description
`pipenv run mypy src/paths tests/tests_paths` passes without errors
 
### Issues Resolved
https://github.com/opensearch-project/opensearch-build/issues/1510
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by peternied — CHANGES_REQUESTED** [review]
Please also update the following to enforce mypy on the new paths you've added:
- .github/workflows/python-tests.yml
- .pre-commit-config.yaml

You can see how this is done in https://github.com/opensearch-project/opensearch-build/pull/1224/files

**Comment by peterzhuamazon** [comment]
The name should be something like `Adding Type Checking ......` not fixing mypy issues. I was quite confused for a while and think there is issue on how to use mypy in our workflows.

File: src/paths/build_output_dir.py [hunk 1]
```diff
@@ -8,5 +8,5 @@
 
 
 class BuildOutputDir(OutputDir):
-    def __init__(cls, filename, cwd=None, makedirs=True):
+    def __init__(cls, filename: str, cwd: str=None, makedirs: bool=True) -> None:
```
Comment (gaiksaya) [inline src/paths/build_output_dir.py:11]: Shouldn't the cwd be of type `Path`?

Comment (abhinavGupta16) [reply src/paths/build_output_dir.py]: Changed it to `Union[str, PathLike[str]]`

Comment (peternied) [reply src/paths/build_output_dir.py]: I think this should stay as a string, in the unit tests the only type we use is string.  In the codebase its only reference is a string in `run_build.py`

**Review by peternied — APPROVED** [review]

File: src/paths/build_output_dir.py (hunk 1 above)
Comment (peternied) [reply src/paths/build_output_dir.py]: Note; the impact of this type is limited, whereas in https://github.com/opensearch-project/opensearch-build/pull/1564 its much broader in scope.

I would still prefer the return type changed to string

Comment (gaiksaya) [reply src/paths/build_output_dir.py]: Makes sense. @abhinavGupta16 are you changing it back to string?

Comment (abhinavGupta16) [reply src/paths/build_output_dir.py]: yes, I changed it all to `str` now

**Review by dblock — APPROVED** [review]

**Pull request merged**